File: cpp/builder/builder.cpp
#include <iostream>
#include <memory>
#include <utility>

class House {
public:
    void add_thing(const std::string &thing) {
        if (!_was_added) {
            _description += " with ";
            _was_added = true;
        } else {
            _description += " and ";
        }
        _description += thing;
    }

    std::string get_description() {
        return _description;
    }

private:
    std::string _description = "House";
    bool _was_added = false;
};

class Builder {
public:
    void add_garage() {
        _house->add_thing("garage");
    }

    void add_garden() {
        _house->add_thing("garden");
    }

    void reset() {
        _house = std::make_shared<House>();
    }

    std::shared_ptr<House> get_result() {
        return _house;
    }

private:
    std::shared_ptr<House> _house = std::make_shared<House>();
};

int main() {
    Builder builder;
    builder.add_garage();
    builder.reset();
    builder.add_garden();
    builder.add_garage();
    std::cout << builder.get_result()->get_description() << std::endl;
    return 0;
}

File: python/builder/builder.py
class House:
    def __init__(self):
        self._description: str = "House"
        self._was_added = False

    def get_description(self) -> str:
        return self._description

    def add_thing(self, thing: str):
        if not self._was_added:
            self._description += " with "
            self._was_added = True
        else:
            self._description += " and "
        self._description += thing


class Builder:
    def __init__(self):
        self._house = House()

    def reset(self):
        self._house = House()

    def add_garage(self):
        self._house.add_thing("garage")

    def add_garden(self):
        self._house.add_thing("garden")

    def get_house(self):
        return self._house


def main():
    builder = Builder()
    builder.add_garage()
    builder.reset()
    builder.add_garden()
    builder.add_garage()
    print(builder.get_house().get_description())


if __name__ == '__main__':
    main()
